src/Kestrel.Core/Internal/HttpConnection.cs

```diff
@@ -143,15 +144,30 @@ public Task StartRequestProcessing<TContext>(IHttpApplication<TContext> applicat
                     adaptedPipelineTask = adaptedPipeline.RunAsync(stream);
                 }
 
-                if (_http1Connection.ConnectionFeatures.Get<ITlsApplicationProtocolFeature>()?.ApplicationProtocol == "h2" &&
-                    Interlocked.CompareExchange(ref _http2ConnectionState, Http2ConnectionStarted, Http2ConnectionNotStarted) == Http2ConnectionNotStarted)
+                var hasTls = _context.ConnectionFeatures.Get<ITlsConnectionFeature>() != null;
+                var applicationProtocol = _context.ConnectionFeatures.Get<ITlsApplicationProtocolFeature>()?.ApplicationProtocol;
+                var http1Enabled = (_context.Protocols & EndPointProtocols.Http1) == EndPointProtocols.Http1;
+                var http2Enabled = (_context.Protocols & EndPointProtocols.Http2) == EndPointProtocols.Http2;
+
+                if (!hasTls && http1Enabled && http2Enabled)
+                {
+                    throw new InvalidOperationException("HTTP/1.x and HTTP/2 cannot both be served on the same cleartext endpoint.");
```

Comment:
Suggest rephrase: "Using both HTTP/1.x and HTTP/2 on the same endpoint requires the use of TLS"